Add torch.linalg.ldl_factor_ex and torch.linalg.ldl_solve

IvanYashchuk opened the pull request:
This PR adds a function for computing the LDL decomposition and a function that can solve systems of linear equations using this decomposition. The result of `torch.linalg.ldl_factor_ex` is in a compact form and it's required to use it only through `torch.linalg.ldl_solve`. In the future, we could provide `ldl_unpack` function that transforms the compact representation into explicit matrices.

Fixes https://github.com/pytorch/pytorch/issues/54847.


cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

IvanYashchuk committed:
Add torch.linalg.ldl_factor_ex and torch.linalg.ldl_solve

This PR adds a function for computing the LDL decomposition and a
function that can solve systems of linear equations using this
decomposition. The result of `torch.linalg.ldl_factor_ex` is in compact
form and it's required to use it only through `torch.linalg.ldl_solve`.
In the future we could provide `ldl_unpack` function that transforms the
compact representation into explicit matrices.

Ref. https://github.com/pytorch/pytorch/issues/54847

IvanYashchuk commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:
```diff
@@ -1247,6 +1321,113 @@ magma_trans_t to_magma(TransposeType trans) {
+  }
+}
+
+void ldl_solve_kernel(
```
It should be possible to add the MAGMA backend for this function as well when using MAGMA 2.6.0
https://bitbucket.org/icl/magma/src/c703d112dcf19eb8c73676cef10888aa2ef73457/ReleaseNotes#lines-48

IvanYashchuk committed:
Use ldl_factor_magma for complex inputs and hermitian=true

IvanYashchuk committed:
Add a todo note about MAGMA backend for ldl_solve

IvanYashchuk committed:
Zero infos before calling ldl_factor_stub

IvanYashchuk committed:
Fix cusolver's ldl_solve version check

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -1011,6 +1146,16 @@ REGISTER_AVX2_DISPATCH(lu_factor_stub, &lu_factor_kernel);
+REGISTER_ARCH_DISPATCH(ldl_factor_stub, DEFAULT, &ldl_factor_kernel);
+REGISTER_AVX512_DISPATCH(ldl_factor_stub, &ldl_factor_kernel);
+REGISTER_AVX2_DISPATCH(ldl_factor_stub, &ldl_factor_kernel);
+REGISTER_VSX_DISPATCH(ldl_factor_stub, &ldl_factor_kernel);
```
Should add `REGISTER_ZVECTOR_DISPATCH` here.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into ldl-factor

IvanYashchuk committed:
Add REGISTER_ZVECTOR_DISPATCH

IvanYashchuk committed:
Add entry for test_namedtuple_return_api

IvanYashchuk committed:
Fix the condition when cusolver 11.1.2+ is not available

IvanYashchuk committed:
Use isIntegralType with includeBool=false

IvanYashchuk committed:
Add correctness test for ldl_factor

IvanYashchuk committed:
Add correctness test for ldl_solve

IvanYashchuk committed:
Do not use f-string

IvanYashchuk committed:
Skip test_ldl_solve for older CUDA versions

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -833,6 +833,141 @@ void triangular_solve_kernel(const Tensor& A, const Tensor& B, bool left, bool u
+  auto leading_dim = A.stride(-1);
+  auto uplo = upper ? 'U' : 'L';
+
+  auto a_stride = A.dim() > 2 ? A.stride(-3) : 0;
```
Nice!

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -833,6 +833,141 @@ void triangular_solve_kernel(const Tensor& A, const Tensor& B, bool left, bool u
+      "PyTorch with LAPACK. Please use PyTorch built with LAPACK support.");
+#else
+  auto batch_size = batchCount(B);
+  if (batchCount(B) == 0 || batchCount(A) == 0 ||
+      batchCount(pivots.unsqueeze(-1)) == 0) {
+    return;
+  }
```
Sa,e as before, this optimisation should go at the top of the `TORCH_IMPL_FUNC`.

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -833,6 +833,141 @@ void triangular_solve_kernel(const Tensor& A, const Tensor& B, bool left, bool u
+
+  auto a_data = A.data_ptr<scalar_t>();
+  auto b_data = B.data_ptr<scalar_t>();
+  auto pivots_ = pivots.to(kInt);
```
In my opinion, we should just accept pivots of type `kInt`.

IvanYashchuk replied:
ldl_factor based on cuSOLVER is returning 32-bit ints, ldl_solve based on cuSOLVER is expecting 64-bit ints. That's why `ldl_solve` allows any integral type that is then transformed to what the library can work with.

Lezcano replied:
Oof. Please leave a comment in the relevant places then.

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -833,6 +833,141 @@ void triangular_solve_kernel(const Tensor& A, const Tensor& B, bool left, bool u
+        ldb,
+        &info);
+  }
+  TORCH_INTERNAL_ASSERT(info == 0);
```
This should probably be a `TORCH_CHECK` throwing a reasonable error, just in case?

IvanYashchuk replied:
It should never throw and if it ever does I'd like the user to submit a bug report, which `TORCH_INTERNAL_ASSERT` conveniently suggests to do.

Lezcano commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:
```diff
@@ -1249,6 +1323,124 @@ magma_trans_t to_magma(TransposeType trans) {
+namespace {
+
+template <typename scalar_t>
+void apply_ldl_factor(
```
```suggestion
void apply_ldl_factor_looped(
```
to go with `ldl_factor_looped_magma`.

IvanYashchuk replied:
I removed "looped" completely, there's no specialized batched variant.
Will add the magma prefix.

Lezcano commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:
```diff
@@ -1249,6 +1323,124 @@ magma_trans_t to_magma(TransposeType trans) {
+  if (factors.is_complex()) {
+    TORCH_CHECK(
+        !hermitian,
+        "torch.linalg.ldl_solve: complex tensors with hermitian=True flag are not supported.");
```
```suggestion
        "torch.linalg.ldl_solve: complex tensors with hermitian=True flag are not supported on CUDA.");
```

Lezcano commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cpp:
```diff
@@ -113,6 +113,180 @@ void lu_solve_batched_cublas(const Tensor& b, const Tensor& lu, const Tensor& pi
+      "PyTorch with cuSOLVER. Please use PyTorch built with cuSOLVER 11.1.2+ (CUDA 11.3.1+) support.");
+#else
+  auto batch_size = batchCount(B);
+  if (batchCount(B) == 0 || batchCount(A) == 0 ||
+      batchCount(pivots.unsqueeze(-1)) == 0) {
+    return;
+  }
```
Move to the main function.

Lezcano commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cpp:
```diff
@@ -113,6 +113,180 @@ void lu_solve_batched_cublas(const Tensor& b, const Tensor& lu, const Tensor& pi
+  auto a_data = A.data_ptr<scalar_t>();
+  auto b_data = B.data_ptr<scalar_t>();
+
+  auto pivots_ = pivots.to(kLong);
```
Same as above.

IvanYashchuk replied:
https://github.com/pytorch/pytorch/pull/69828#discussion_r786428218

Lezcano commented on test/test_linalg.py:
```diff
@@ -8336,6 +8337,112 @@ def test_tensordot(self, device):
+        def run_test(shape, batch, hermitian, upper):
+            A = random_hermitian_pd_matrix(shape, *batch, dtype=dtype, device=device)
+            actual_factors, actual_pivots, info = torch.linalg.ldl_factor_ex(A, hermitian=hermitian, upper=upper)
+            tri = torch.triu if upper else torch.tril
+            diag = 1 if upper else -1
+            actual_L = tri(actual_factors, diagonal=diag)
```
A bit tricky to read. Perhaps a ternary operator is a better option here.

IvanYashchuk replied:
`if ... else ...` is already used here to define `tri` and `diag`. Do you want this piece of code to be modified to
```py
actual_L = torch.triu(actual_factors, diagonal=diag) if upper else torch.tril(actual_factors, diagonal=diag)
```

Lezcano replied:
or a fully fledged  if-else if the ternary operator is too cluttered ternary operator, whatever you see fit. Either
`actual_L = torch.triu(actual_factors, diagonal=1) if upper else torch.tril(actual_factors, diagonal=-1)`
or 
```python
if upper:
  actual_L = torch.triu(actual_factors, diagonal=1)
else:
  actual_L = torch.tril(actual_factors, diagonal=-1)
```
Probably the if-else is more readable?

Lezcano commented on test/test_linalg.py:
```diff
@@ -8336,6 +8337,112 @@ def test_tensordot(self, device):
+        magma_254_available = self.device_type == 'cuda' and _get_magma_version() >= (2, 5, 4)
+        hermitians = (True, False) if dtype.is_complex and magma_254_available else (False,)
+
+        shapes = (5,)
```
Add a test for the zero shape as well or nah?

IvanYashchuk replied:
I had it initially but then it requires `if numel == 0` because there are row permutations before comparison.

It's tested in OpInfo samples.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,102 @@
+ldl_factor_ex = _add_docstr(_linalg.linalg_ldl_factor_ex, r"""
+linalg.ldl_factor_ex(A, *, upper=False, hermitian=False, check_errors=False, out=None) -> (Tensor, Tensor, Tensor)
+
+Computes the LDL decomposition of a Hermitian/symmetric indefinite matrix.
+When :attr:`A` is complex valued it can be Hermitian (:attr:`hermitian` = `True`)
+or symmetric (:attr:`hermitian` = `False`).
```
The first paragraph should be just one line for it to be formatted properly in the linear algebra docs page.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,102 @@
+linalg.ldl_factor_ex(A, *, upper=False, hermitian=False, check_errors=False, out=None) -> (Tensor, Tensor, Tensor)
+
+Computes the LDL decomposition of a Hermitian/symmetric indefinite matrix.
+When :attr:`A` is complex valued it can be Hermitian (:attr:`hermitian` = `True`)
```
It should be ``:attr:` hermitian`\ `= True`` for the equals to be formatted "properly", as annoying as this is. Same below.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,102 @@
+ldl_factor_ex = _add_docstr(_linalg.linalg_ldl_factor_ex, r"""
+linalg.ldl_factor_ex(A, *, upper=False, hermitian=False, check_errors=False, out=None) -> (Tensor, Tensor, Tensor)
+
+Computes the LDL decomposition of a Hermitian/symmetric indefinite matrix.
```
```suggestion
Computes the LDL decomposition of a Hermitian/symmetric matrix.
```

It'd be good to have a mathematical definition of this function somewhere in the description, similar to that in SVD
https://pytorch.org/docs/stable/generated/torch.linalg.svd.html
Even better, this should be in a potential `linalg.ldl` function, and here we should refer to the docs of this other function.

IvanYashchuk replied:
SVD is easy to describe because the return is exactly the matrices that define the decomposition. Here the decomposition is in compact form. We don't have this `linalg.ldl` function yet, so we shouldn't refer to it. This function does refer to `linalg.ldl_solve` which is created to work with the compact representation this function returns. I think it's rather difficult and unnecessary to describe the details of the representation. It's enough to mention that it's supposed to be used with `ldl_solve`.

IvanYashchuk replied:
Is your suggestion to remove the "indefinite" word in this sentence?

Lezcano replied:
Re the definite, yes.

About that, sure thing. I think it would be good to follow a description as similar similar to that in https://pytorch.org/docs/master/generated/torch.linalg.lu_factor.html as possible for consistency.
If describing the format of the decomposition is too difficult, I'd say we put in its place a link to the relevant LAPACK page where it's described.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,102 @@
+
+Computes the LDL decomposition of a Hermitian/symmetric indefinite matrix.
+When :attr:`A` is complex valued it can be Hermitian (:attr:`hermitian` = `True`)
+or symmetric (:attr:`hermitian` = `False`).
```
```suggestion
or symmetric (:attr:`hermitian` = `False`). If :attr:`A` is real, it is assumed to be a symmetric matrix.
```

IvanYashchuk replied:
What does "assumed" mean here? Do you want to have mentions in the docs that only lower/upper triangular part of the inputs is used in the computation?

Lezcano replied:
What I meant to convey here is that in the real case the hermitian attribute does not change anything. You can add to the proposed sentence "irrespective of the value of ``:attr:`hermitian` ``" if you think that makes it clearer.

So, if we go with the user of the `lower` argument (which I'd prefer not to) we would need to describe exactly how this function works. In particular we'd need to describe those nuances, and probably say that the gradient is upper/lower triangular. If the inputs are assumed to be symmetric / Hermitian (no `lower` argument), then we can skip all this.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,102 @@
+:math:`L D L^T` (:attr:`upper` = `False`) or :math:`U D U^T` (:attr:`upper` = `True`).
+If :attr:`hermitian` is `True` then transpose operation is the conjugate transpose.
+
+:math:`L` (or :math:`U`) and :math:`D` are stored in compact form in ``factors``.
+This format is used by :func:`torch.linalg.ldl_solve` for solving linear systems.
```
Add a link to the LAPACK docs where the format is described. Same for the `info` below.

IvanYashchuk replied:
I'm hesitant to use links to LAPACK documentation because URL doesn't look like something that persists in history and not changing between the releases:
https://www.netlib.org/lapack/explore-html/d3/db6/group__double_s_ycomputational_gad91bde1212277b3e909eb6af7f64858a.html

Lezcano replied:
Fair enough. Then perhaps say "the returned is that of LAPACK's sytrf" or so?

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,102 @@
+Also supports batches of matrices, and if :attr:`A` is a batch of matrices then
+the output has the same batch dimensions.
+
+.. note:: If :attr:`A` is on a CUDA device, this function may synchronize that device with the CPU.
```
Use `common_notes["sync_note_ex"]`.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,102 @@
+
+.. note:: If :attr:`A` is on a CUDA device, this function may synchronize that device with the CPU.
+
+.. warning:: This function is "experimental" and it may change in a future PyTorch release.
```
Use `common_notes["experimental_warning"]`.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,102 @@
+Examples::
+
+    >>> A = torch.randn(3, 3)
+    >>> A
```
Make `A` symmetric, otherwise this example might look very weird.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,102 @@
+linalg.ldl_solve(factors, pivots, B, *, upper=False, hermitian=False, out=None) -> Tensor
+
+Computes the solution of a system of linear equations using the LDL factorization.
+
```
I would add a pararaph describing this using LaTeX in matheamtical notation, as: It solves `LDLX = B`.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,102 @@
+Also supports batches of matrices, and if :attr:`A` is a batch of matrices then
+the output has the same batch dimensions.
+
+.. warning:: This function is "experimental" and it may change in a future PyTorch release.
```
Same as above.

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4953,6 +4953,45 @@ def sample_inputs_linalg_cholesky_inverse(op_info, device, dtype, requires_grad=
+        factors.requires_grad = requires_grad
+        for B_batch_shape in ((), factors.shape[:-2]):
+            B = make_tensor((*B_batch_shape, factors.shape[-1], S), device=device, dtype=dtype, requires_grad=requires_grad)
+            out.append(SampleInput(factors, args=(pivots, B), kwargs=dict(hermitian=True)))
```
You can use `yield` if you prefer now. It hsould be more efficient for some opinfos.

IvanYashchuk replied:
Ok, I've changed to `yield`, but it doesn't affect the runtime of `ldl_solve` tests in any way.

lezcano reviewed (commented):
As major points, I think it'd also be good to add a `linalg.ldl_factor` function.

I would strongly suggest to remove the `upper` keyword, as
1. It's has very few applications.
2. The derivative will look very awkward with it, as the gradients would need to be upper-triangular, rather than symmetric.
At the moment, even though we support this in `linalg.cholesky`, if you try to use it with an upper-triangular matrix in place of the SPD matrix, the gradients will be incorrect.

IvanYashchuk commented on test/test_namedtuple_return_api.py:
```diff
@@ -87,6 +88,7 @@ def test_namedtuple_return(self):
             op(operators=['linalg_inv_ex'], input=(), names=('inverse', 'info'), hasout=True),
             op(operators=['linalg_lu_factor'], input=(), names=('LU', 'pivots'), hasout=True),
             op(operators=['linalg_lu_factor_ex'], input=(), names=('LU', 'pivots', 'info'), hasout=True),
+            op(operators=['linalg_ldl_factor_ex'], input=(), names=('factors', 'pivots', 'info'), hasout=True),
```
But then it's LDU! (because it contains matrix D, and either matrix L or U).

Lezcano replied:
As mentioned below, I don't think we should allow for `upper=True`, but regardless, I'd say this is the `LDL` factor for the same reason this is the `LDL` decomposition and not the `LDU` decomposition.

IvanYashchuk replied:
Okay, if we drop `upper` we don't have the "U" component. Then this return should probably be named `LD` and not `LDL`. It contains matrices L and D.
I feel that "factors" is more generic and less opinionated. Then once we have eventually `ldl(A)` we can use return names separate for each matrix `ldl(A) -> (L, D, pivots)`.

Lezcano replied:
I think `LD` would be nice. I think it'd be better for consistency with `cholesky` which returns `L` and `lu_factor` whcih returns `LU`.

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -206,6 +328,67 @@ extern "C" void strsm_(char *side, char *uplo, char *trans, char *diag, int *n,
+ const Tensor& pivots,
+ const Tensor& B,
+ bool upper,
+ bool hermitian) {
```
This flag is cheap in BLAS (hence solve_triangular). But ldl_solve LAPACK doesn't even have transpose flags and doesn't have left/right flags.

Lezcano replied:
What I outlined up there is a way to implement it at the `TORCH_IMPL_FUNC` level via a tad of extra preprocessing and postprocessing.

IvanYashchuk replied:
I'd rather submit an issue after this PR because of "extra preprocessing and postprocessing". It could be done in the future hypothetical generic `factorize + solve` functions that wrap these factorization-specific solvers.

It's possible to pass a triangular matrix to `ldl_factor` and everything would still work as expected and we can't restrict the input to be strictly symmetric/Hermitian. Since passing triangular matrix is allowed then `A != A.mT` and it's not possible to reuse the same factorization for the "left" solver.

Lezcano replied:
So, in my opinion, this function should expect the inputs to be always symmetric / Hermitian according to the flag, and thus, its gradients should also be symmetric or Hermitian.

If that's the case, we should be allowed to do the thing proposed above although, of course, it needn't be in this PR. Could you leave a comment at the top of the function sayin that it'd be cool to implement that?

Now, about supporting triangular matrices, if you're keen on doing so, fine (although I'd prefer not to do so) but then you'd need to:
1. Document this properly, as it's not the expected behaviour and it's currently undocumented.
2. Be careful with the implementation of the gradients. This function will oddly return upper/lower triangular gradients, as opposed to what `linalg.cholesky` and `linalg.eigh` do. We should probably also document this, as it's a bit odd.

IvanYashchuk replied:
Both `linalg.cholesky` and `linalg.eigh` work with triangular input:
```
In [1]: import torch

In [2]: a = torch.randn(3, 3)

In [3]: a = a @ a.mT

In [4]: torch.allclose(torch.linalg.cholesky(a), torch.linalg.cholesky(a.tril()))
Out[4]: True

In [5]: [torch.allclose(x, y) for x, y in zip(torch.linalg.eigh(a), torch.linalg.eigh(a.tril()))]
Out[5]: [True, True]
```

Lezcano replied:
If you pass a triangular output the gradient and forward AD will be incorrect

Lezcano replied:
Note that this was one of the points raised that led to the removal of the `upper` argument in `eigh` and `eigvalsh` in the array API https://github.com/data-apis/array-api/issues/217

IvanYashchuk replied:
For Cholesky and eigh it is symmetrized and correct. Previously PyTorch (pre ~1.5) would return triangular derivatives for Cholesky, but it was changed.

```python
In [1]: import torch

In [2]: a = torch.randn(3, 3)

In [3]: a = a @ a.mT

In [4]: a_tril = a.tril()

In [5]: g = torch.randn(3, 3)

In [6]: a_tril.requires_grad_(True)
Out[6]:
tensor([[ 9.1411,  0.0000,  0.0000],
        [-6.0545,  4.1667,  0.0000],
        [-3.7081,  2.3873,  3.4195]], requires_grad=True)

In [7]: L = torch.linalg.cholesky(a_tril)

In [8]: L
Out[8]:
tensor([[ 3.0234,  0.0000,  0.0000],
        [-2.0025,  0.3957,  0.0000],
        [-1.2264, -0.1736,  1.3730]], grad_fn=<LinalgCholeskyExBackward0>)

In [9]: L.backward(g)

In [10]: a_tril.grad
Out[10]:
tensor([[ 0.0103,  0.9693, -0.4156],
        [ 0.9693,  1.7575, -0.7359],
        [-0.4156, -0.7359,  0.6651]])
```

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -3847,6 +4314,139 @@ Tensor _det_lu_based_helper_backward_helper(
+  at::native::resize_output(info, info_shape);
+
+  auto info_ = info.expect_contiguous();
+  info_->zero_();
```
It's here because it doesn't work as expected 🥲 
Well, I can wrap it in `if (self.numel() == 0)`.

Lezcano replied:
Then, if we move the `self.numel()` check to the beginning of this function we should be good, right? See https://github.com/pytorch/pytorch/pull/69827

IvanYashchuk commented:
> I would strongly suggest to remove the `upper` keyword, as
> 
> 1. It's has very few applications.
> 2. The derivative will look very awkward with it, as the gradients would need to be upper-triangular, rather than symmetric.
>    At the moment, even though we support this in `linalg.cholesky`, if you try to use it with an upper-triangular matrix in place of the SPD matrix, the gradients will be incorrect.

It doesn't make sense to leave `upper` when the intended use case is ldl_factor + ldl_solve. But the application of this kwarg is the same as in the case of `cholesky` - to get lower or upper triangular matrix output when working with the factors explicitly. So I'd say it mainly controls the output ldl_factor and considering the upper/lower part of the input is accidental.

We can of course remove it now, and maybe add it back later in case someone needs it.

Lezcano commented:
Note that `UDU = A` is equivalent to `A^T = LDL`, and since `A=A^T`, `U=L^T`. Same applies when `hermitian=False` but doing `L^H`.

I do not know of any benefit of using upper over lower triangular, so I don't know why would anyone want to use one over the other, even less so now that we have a nice `solve_triangular` that can solve equations on the left / right for upper / lower triangular matrices.

We discussed leaving `upper=False` in `linalg.cholesky` as a compatibility argument, but deprecate it otherwise, but we never followed through with that plan cc @mruberry wdyt

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -833,6 +833,141 @@ void triangular_solve_kernel(const Tensor& A, const Tensor& B, bool left, bool u
+      "PyTorch with LAPACK. Please use PyTorch built with LAPACK support.");
+#else
+  auto batch_size = batchCount(A);
+  if (batch_size == 0) {
+    return;
+  }
```
This is not an optimization, it's used for safety to avoid reference null pointers.
I will modify it to make it common for all implementations.

mruberry commented:
> Note that `UDU = A` is equivalent to `A^T = LDL`, and since `A=A^T`, `U=L^T`. Same applies when `hermitian=False` but doing `L^H`.
> 
> I do not know of any benefit of using upper over lower triangular, so I don't know why would anyone want to use one over the other, even less so now that we have a nice `solve_triangular` that can solve equations on the left / right for upper / lower triangular matrices.
> 
> We discussed leaving `upper=False` in `linalg.cholesky` as a compatibility argument, but deprecate it otherwise, but we never followed through with that plan cc @mruberry wdyt

It's easier not to include the `upper` kwarg and see if we get requests for it, but if it's easy to implement it seems harmless. I'd let @IvanYashchuk decide.

IvanYashchuk committed:
Use at::native::contiguous_strides

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -206,6 +328,67 @@ extern "C" void strsm_(char *side, char *uplo, char *trans, char *diag, int *n,
+      B.dim(),
+      " dimensions instead");
+  TORCH_CHECK(
+      at::isIntegralType(pivots.scalar_type(), /*includeBool=*/false),
```
No, LAPACK requires kInt because we use 32-bit interface, cuSOLVER requires 64-bit (kLong), so we'll allow any integer that will be converted either to kInt or kLong.
https://github.com/pytorch/pytorch/pull/69828#discussion_r786428218.

IvanYashchuk committed:
Remove unnecessary resizes and checks for shape

IvanYashchuk committed:
Use at::native::borrow_else_clone

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -3847,6 +4314,139 @@ Tensor _det_lu_based_helper_backward_helper(
+ const Tensor& B,
+ bool upper,
+ bool hermitian,
+ const Tensor& result) {
```
I have troubles naming A and B other than that, I could use `input` and `other` here. But `X` is something that is unknown not the result. I don't think conciseness is what we're after when we name arguments.

Lezcano replied:
My point is that this function solves a linear system, which is often written as `AX = B`. As such, the result of this function is a matrix (or batch) almost always denoted as `X` in maths, hence my suggestion.

Now, if there's no standard name for the result of a function, `result`, `out` or any other name is perfectly fine, of course.

IvanYashchuk committed:
Add a note on integer type of pivots for ldl_solve

IvanYashchuk committed:
The first paragraph should be just one line for it to be formatted properly in the linear algebra docs page

IvanYashchuk committed:
Use symmetric matrix in ldl_factor example

IvanYashchuk committed:
Use sync_note_ex and experimental_warning notes

IvanYashchuk committed:
Mention in the docs that ldl decomposition is in the compact form

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into ldl-factor

IvanYashchuk committed:
Use at::_linalg_check_errors

IvanYashchuk committed:
Prevent invalid memory reads when input has 0s in batch dims

IvanYashchuk committed:
Remove scalar_type checks, it's done by structured kernels from meta function implementation

IvanYashchuk committed:
Omit 'looped' in function names

IvanYashchuk committed:
Update error message for complex inputs with hermitian=True on CUDA

IvanYashchuk committed:
Update docs formatting

IvanYashchuk committed:
Update documentation

Use 'factorization' term
Add a note that hermitian upper flag should match in ldl_factor and
ldl_solve

IvanYashchuk committed:
clang-format

IvanYashchuk committed:
Use yield for sample_inputs_linalg_ldl_solve

IvanYashchuk committed:
Rename 'factors' -> 'LD'

IvanYashchuk committed:
Remove 'upper' keyword argument from public interface

IvanYashchuk commented:
Thank you, Mario, for your review! I think I've addressed most of the points, a few highlights of the changes (and things I didn't do):
* I removed a few unnecessary resizes and tests still pass. It's good also that at::native::resize_output does nothing actually if shapes match, but not using it at all is better.
* I didn't rename `result` -> `X` (https://github.com/pytorch/pytorch/pull/69828#discussion_r786264727)
* I didn't add a link to LAPACK documentation, because I fear it might break in the future (https://github.com/pytorch/pytorch/pull/69828#discussion_r792641475)
* I didn't add a LaTeX description of what "compute solution of a system of linear equations" means (https://github.com/pytorch/pytorch/pull/69828#discussion_r786276379)
* I renamed `factors` -> `LD` (https://github.com/pytorch/pytorch/pull/69828#discussion_r786507242)
* I removed the `upper=False` keyword argument from `ldl_factor_ex` and `ldl_solve functions`

Lezcano commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:
```diff
@@ -1249,6 +1323,124 @@ magma_trans_t to_magma(TransposeType trans) {
+  auto pivots_stride = pivots.dim() > 1 ? pivots.stride(-2) : 0;
+
+  auto a_data = A.data_ptr<scalar_t>();
+  Tensor pivots_cpu =
+      at::empty_like(pivots, pivots.options().device(kCPU).pinned_memory(true));
+  auto pivots_data = pivots_cpu.data_ptr<magma_int_t>();
+  Tensor info_cpu =
+      at::empty_like(info, info.options().device(kCPU).pinned_memory(true));
+  auto info_data = info_cpu.data_ptr<magma_int_t>();
```
We don't need to create all the tensors. We can simply reserve enough memory for the pivot of one matrix (`std::vector` should do for that) and then copy them to the right position in pivots and info in each iteration of the loop (flattening the few dimensions via a `view`, for example)

IvanYashchuk replied:
I prefer using `at::Tensor` for reserving contiguous blocks of memory on the CPU.

IvanYashchuk replied:
I prefer using `at::Tensor` for reserving contiguous blocks of memory on the CPU instead of `std::vector` and we need it to be pinned for performance.
I don't think it's necessarily faster to reserve memory for one matrix/pivot in a batch and copy it one-by-one back to GPU when the result is computed.

Lezcano replied:
This is what we're currently doing in `cusovler_gesvd`, so we should figure out what's the best way of doing this and do it consistently across the code base.
A priori, I'd say that if they are in pinned memory and all these things are async (the copies can be async, as long as you synchronise at the end) it should be better not to reserve the full batch of matrices but just the one matrix of the correct size but idk. It's not that important really.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,109 @@
+If :attr:`hermitian` is `True` then transpose operation is the conjugate transpose.
+
+:math:`L` (or :math:`U`) and :math:`D` are stored in compact form in ``LD``.
+This format is used by :func:`torch.linalg.ldl_solve` for solving linear systems.
```
Nit.
```suggestion
They follow the format specified by LAPACK's `sytrf' function. 
These tensors may be used in :func:`torch.linalg.ldl_solve` to solve linear systems.
```

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -813,6 +813,109 @@
+""" + r"""
+
+Args:
+    A (Tensor): the Hermitian/symmetric `n \times n` matrix or the batch of such matrices of size
```
```suggestion
    A (Tensor): tensor of shape (*, n, n) where * is zero or more batch dimensions consisting of symmetric or Hermitian matrices.
```

lezcano reviewed (commented):
I left a few very minor comments. There's nothing too important.

The only thing that I think it'd be good to have is the `linalg.ldl_factor` function that simply calls into `linalg.ldl_factor_ex`. I would also put the docs you have written into this function, and write some very short docs for `ldl_factor_ex` referencing those. See https://pytorch.org/docs/master/generated/torch.linalg.lu_factor_ex.html

Also, I believe you forgot to add the doc entries to the `linalg.rst` file.

IvanYashchuk committed:
Check pivots and LD are of compatible shapes

IvanYashchuk committed:
Add _magma and _cusolver suffixes

IvanYashchuk committed:
Apply Mario's nit

IvanYashchuk committed:
Fix test_fx failures

IvanYashchuk committed:
Enable hermitian=True testing for CPU

IvanYashchuk committed:
Skip hermitian=True sample inputs on CUDA withput MAGMA 2.5.4+

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -3774,6 +4237,95 @@ Tensor _det_lu_based_helper_backward_helper(
+ const Tensor& B,
+ bool hermitian,
+ const Tensor& result) {
+  if (LD.numel() == 0 || pivots.numel() == 0) {
```
No, it's not allowed.

Lezcano replied:
The it might be best to simply do `if (LD.numel() == 0)` for readability. Now, it's not particularly important

IvanYashchuk committed:
Update torch/linalg/__init__.py

Co-authored-by: Lezcano <Lezcano@users.noreply.github.com>

IvanYashchuk committed:
Add entries to linalg.rst

IvanYashchuk commented:
periodic-linux-bionic-cuda11.5 fails with 
 ```
 ❌ Failure: TestCommonCUDA.test_dtypes_linalg_ldl_solve_cuda
 torch.complex128 : Attempted to compare [set] types: Expected: {torch.float64, torch.float32, torch.complex64}; Actual: {torch.float64, torch.float32, torch.complex64, torch.complex128}.
The supported dtypes for linalg.ldl_solve on cuda according to its OpInfo are
        {torch.float64, torch.float32, torch.complex64, torch.complex128}, but the detected supported dtypes are {torch.float64, torch.float32, torch.complex64}.
        The following dtypes should be removed from the OpInfo: {torch.complex128}.
 ```

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -3774,6 +4244,95 @@ Tensor _det_lu_based_helper_backward_helper(
+    return;
+  }
+
+  Tensor B_broadcast;
+  std::tie(B_broadcast, std::ignore) =
+      _linalg_broadcast_batch_dims(B, LD, /*don't check errors*/ nullptr);
```
Is this now just equivalent to `auto B_broadcast = B.expand_as(result);`?

IvanYashchuk replied:
Right. I removed it now because we don't need to expand `B` as `result` to call `result.copy_(B)`.

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -3774,6 +4244,95 @@ Tensor _det_lu_based_helper_backward_helper(
+  auto pivots_ = pivots.expect_contiguous();
+
+  auto LD_ = at::native::borrow_else_clone(
+      result.mT().is_contiguous(), LD, LD, /*row_major=*/false);
```
```suggestion
      LD.mT().is_contiguous(), LD, LD, /*row_major=*/false);
```

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into ldl-factor

IvanYashchuk committed:
Rename factors -> LD

IvanYashchuk committed:
Add lazy cuda kernels

IvanYashchuk committed:
Add ldl_factor (not returning info) variant

IvanYashchuk committed:
Add docs and OpInfo for ldl_factor

IvanYashchuk committed:
Update aten/src/ATen/native/BatchLinearAlgebra.cpp

Co-authored-by: Lezcano <Lezcano@users.noreply.github.com>

IvanYashchuk committed:
Remove expanding B in ldl_solve_out because copy_ can handle broadcasting

IvanYashchuk commented:
I added the `ldl_factor` variant that doesn't return `info` to this PR.

IvanYashchuk committed:
Add ldl_factor to linalg.rst

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -4310,6 +4310,30 @@ TORCH_IMPL_FUNC(linalg_ldl_factor_ex_out)
+    bool hermitian) {
+  Tensor LD, pivots, info;
+  std::tie(LD, pivots, info) =
+      at::linalg_ldl_factor_ex(self, hermitian, /*check_errors=*/false);
```
This should be `chech_errors=True` by default. At least that's what we do in `lu_factor`.

IvanYashchuk replied:
It's `false` in `lu_factor`. It's done so that the error message has the function name `ldl_factor` and not `ldl_factor_ex`.

Lezcano replied:
You are completely right. Even though I wrote that function a couple months ago I totally misremembered this.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -833,19 +833,62 @@
+A positive integer indicates the diagonal element of :math:`D` that is zero.
+Division by 0 will occur if the result is used for solving a system of linear equations.
+``info`` filled with zeros indicates that the factorization was successful.
+If ``check_errors=True`` and ``info`` contains positive integers, then a RuntimeError is thrown.
```
Nit. ``RuntimeError`` between backtiks.

IvanYashchuk replied:
```suggestion
If ``check_errors=True`` and ``info`` contains positive integers, then a `RuntimeError` is thrown.
```

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -833,19 +833,62 @@
+If ``check_errors=True`` and ``info`` contains positive integers, then a RuntimeError is thrown.
+
 """ + fr"""
 .. note:: {common_notes["sync_note_ex"]}
```
This should be
```
""" + fr"""
.. note:: {common_notes["sync_note_ex"]}
.. warning:: {common_notes["experimental_warning"]}
""" + r"""
```

IvanYashchuk replied:
It's there on line 895. The diff is hidden.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -833,19 +833,62 @@
+""" + fr"""
+.. note:: {common_notes["sync_note_has_ex"].format("torch.linalg.ldl_factor_ex")}
+
+.. warning:: {common_notes["experimental_warning"]}
```
This line should be in the `_ex` variant.

IvanYashchuk replied:
It's a new function (`torch.linalg.ldl_factor`) that doesn't exist in NumPy, SciPy, and Array API. So I thought it would be fine to have this note:
> This function is "experimental" and it may change in a future PyTorch release.

Lezcano replied:
IMO, this function is very sound and it's here to stay, so I think it's fine to skip this note and just put it in the `_ex` variant.

IvanYashchuk replied:
Okay, I'll remove it.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -821,8 +821,8 @@
+ldl_factor = _add_docstr(_linalg.linalg_ldl_factor, r"""
+linalg.ldl_factor(A, *, hermitian=False, out=None) -> (Tensor, Tensor)
 
 Computes a compact representation of the LDL factorization of a Hermitian or symmetric indefinite matrix.
```
```suggestion
Computes a compact representation of the LDL factorization of a Hermitian or symmetric (possibly indefinite) matrix.
```
either that or simply delete the "indefinite", as this function also works when the input matrix is positive/negative definite.

lezcano reviewed (commented):
Nice simplifications! I left a few notes on the docs, and also one on the default behaviour on the non-ex function.

IvanYashchuk committed:
Update torch/linalg/__init__.py

Co-authored-by: Lezcano <Lezcano@users.noreply.github.com>

IvanYashchuk commented on torch/linalg/__init__.py:
```diff
@@ -833,19 +833,62 @@
+A positive integer indicates the diagonal element of :math:`D` that is zero.
+Division by 0 will occur if the result is used for solving a system of linear equations.
+``info`` filled with zeros indicates that the factorization was successful.
+If ``check_errors=True`` and ``info`` contains positive integers, then a RuntimeError is thrown.
```
```suggestion
If ``check_errors=True`` and ``info`` contains positive integers, then a `RuntimeError` is thrown.
```

IvanYashchuk committed:
Update torch/linalg/__init__.py

Co-authored-by: Lezcano <Lezcano@users.noreply.github.com>

IvanYashchuk committed:
Remove experimental_warning from ldl_factor docs

IvanYashchuk committed:
Add namedtuple test for ldl_factor

IvanYashchuk committed:
Add overrides entry for ldl_factor

lezcano reviewed (approved):
Thank you for the fantastic PR @IvanYashchuk !

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into ldl-factor

IvanYashchuk commented:
@mruberry, could you please review this PR?

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into ldl-factor

mruberry reviewed (approved):
Let's rebase and get the tests to pass and then merge this as-is. Sorry for the delay, @IvanYashchuk

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into ldl-factor

IvanYashchuk commented:
Resolved the conflict, let's see if CI is green.

IvanYashchuk commented:
Test failure is real:
```py
FAIL [0.030s]: test_out_linalg_ldl_solve_cpu (__main__.TestCommonCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 808, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 808, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 808, in dep_fn
    return fn(slf, *args, **kwargs)
  [Previous line repeated 1 more time]
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 773, in test_wrapper
    return test(*args, **kwargs)
  File "test_ops.py", line 526, in test_out
    _compare_out(_case_one_transform)
  File "test_ops.py", line 495, in _compare_out
    self.assertEqual(expected, out)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2211, in assertEqual
    msg=msg,
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_comparison.py", line 1074, in assert_equal
    raise error_metas[0].to_error()
AssertionError: Tensor-likes are not close!

Mismatched elements: 25 / 50 (50.0%)
Greatest absolute difference: 20.296348676085472 at index (1, 4, 2) (up to 1e-05 allowed)
Greatest relative difference: 100.66243389723535 at index (1, 4, 2) (up to 1.3e-06 allowed)
```

IvanYashchuk committed:
Fixed out variant when B tensor is broadcast along batch dimensions

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into ldl-factor

IvanYashchuk committed:
Make test_correct_module_names pass

IvanYashchuk committed:
Revert "Make test_correct_module_names pass"

This reverts commit 17dea1fbaf3fff5b2221aa0cec4ba3f692461e41.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into ldl-factor

IvanYashchuk committed:
Revert "Revert "Make test_correct_module_names pass""

This reverts commit 66c3ecca49ce55bda2a2afab016658a659572024.

IvanYashchuk commented:
With the recent changes `test_correct_module_names` from `test/test_public_bindings.py` fails with:
```py
# torch.linalg.ldl_factor:
  - Is public: it is an attribute that does not start with `_` on a module that does not have `__all__` defined
  - Does NOT look public: because its `__module__` attribute (`torch._C._linalg`) is not within the torch library or does not start with the submodule where it is defined (`torch.linalg`)
  - You can do either of these two things to fix this problem:
    - To make it NOT public: either define a `__all__` for `torch.linalg` or add a `_` at the beginning of the name
    - To make it look public: make sure the `__module__` is properly set and points to a submodule of `torch.linalg`
```
Defining `__module__` as`"torch.linalg"` makes the other test fail from `test/test_fx_experimental.py` (`test_normalize_operator_exhaustive`) because `torch.fx.symbolic_trace` behaves differently.

Before defining `ldl_factor.__module__ = "torch.linalg"` the FX symbolic trace is:
```py
def forward(self, arg_0):
    linalg_ldl_factor = torch._C._linalg.linalg_ldl_factor(arg_0, hermitian = False);  arg_0 = None
    return linalg_ldl_factor
```
After setting `ldl_factor.__module__ = "torch.linalg"` (which doesn't work)
```py
def forward(self, arg_0):
    linalg_ldl_factor = torch.linalg.linalg_ldl_factor(arg_0, hermitian = False);  arg_0 = None
    return linalg_ldl_factor
```

We can't change the `__name__` attribute to drop the `linalg_` prefix in the name. I think in this situation it's fine to exceptionally add new entries to `allowlist_for_publicAPI.json` and then deal with the whole `torch.linalg` module at once later. @albanD, what do you think?

albanD commented:
If the `__name__` property is completely wrong, that is an issue I think...
> We can't change the __name__ attribute to drop the linalg_ prefix in the name.

Why not?

>  I think in this situation it's fine to exceptionally add new entries to allowlist_for_publicAPI.json and then deal with the whole torch.linalg module at once later. @albanD, what do you think?

I wouldn't say it is an ok thing to do. But I won't block the PR on that.

IvanYashchuk committed:
Skip ldl for fx test

IvanYashchuk committed:
Move setting __module__ to the end of the file

IvanYashchuk committed:
Revert "Move setting __module__ to the end of the file"

This reverts commit 2a9e70dad8a44e011d4f8eca71028a5b64d20cf2.

IvanYashchuk committed:
Revert "Skip ldl for fx test"

This reverts commit 62298e12a0e85708c630416f37800c64aba66259.

IvanYashchuk committed:
Revert "Revert "Revert "Make test_correct_module_names pass"""

This reverts commit ea51ffaeebf278f677042cbb13908edad3dc0752.

IvanYashchuk committed:
Make test_public_bindings pass

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into ldl-factor

IvanYashchuk committed:
Skip ROCm because CI has older version of MAGMA than 2.5.4

IvanYashchuk committed:
Merge branch 'master' into ldl-factor

IvanYashchuk commented:
@pytorchbot merge this

albanD commented:
Note that "lint failed" here means that it didn't finish yet :D

albanD commented:
cc @janeyx99 maybe the fix from https://github.com/pytorch/pytorch/pull/74887 is not working properly to call out that the test didn't finish yet instead of failed?

janeyx99 commented:
@alban I will attempt a fix for that

IvanYashchuk commented:
@pytorchbot merge this

janeyx99 commented:
> cc @janeyx99 maybe the fix from https://github.com/pytorch/pytorch/pull/74887 is not working properly to call out that the test didn't finish yet instead of failed?

Fix attempted https://github.com/pytorch/pytorch/pull/76559 just for future ref
